src/librustc/dep_graph/graph.rs

```diff
@@ -697,7 +778,14 @@ impl DepGraph {
         }
 
         // ... emitting any stored diagnostic ...
-        {
+        if did_allocation {
+            // Only the thread which did the allocation emits the error messages
+
+            // FIXME: Ensure that these are printed before returning for all threads.
```

Comment:
I left this `FIXME` as future work and added it to https://github.com/rust-lang/rust/issues/48685